solutions/gold/rmi-17-Hangman2.mdx

```diff
@@ -0,0 +1,154 @@
+cases given in the hint.
+
+### Case 1: $n < k$
+Because $n < \sqrt{nk}$, we can brute force each pair of strings and check if
+they differ by no more than two characters. This takes $\mathcal{O}(N^2K)$ time,
+which is equivalent to $\mathcal{O}(NK\sqrt{NK})$ time.
+
+### Case 2: $n \geq k$
+Let's compare two strings. Note how allowing at most two characters to be
+different is equivalent to ignoring two characters in both strings and checking
+for equality. Thus, we can brute force on every pair of characters that we are
+ignoring, and then check for equality between strings.
+
+Let the pair of characters we are brute forcing be $(i, j)$. Then,
+we have the following intervals that need to be identical in both strings:
+
+$[0, i - 1], [i + 1, j - 1], [j + 1, k - 1]$
+
+Now, if we want to compare our two strings, then we just need the substrings to
+be equal. This can be checked using string hashing.
+
+For each string in our list, we need to find another string that has the same
+hashes. This is best done by sorting all the tuples of hashes, and then checking
+adjacent items and seeing if the hashes are equal.
+
+Overall, this takes $\mathcal{O}(NK^2\log({N}))$ time. With a bit more analysis,
+you can reach the time complexity of $\mathcal{O}(NK\sqrt{NK}\log(\sqrt{NK}))$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(NK\sqrt{NK}\log(\sqrt{NK}))$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+// BeginCodeSnip{String Hashing Template (from the module)}
+class HashedString {
+  private:
+	// change M and B if you want
+	static const long long M = 1e9 + 9;
+	static const long long B = 9973;
+
+	// pow[i] contains B^i % M
+	static vector<long long> pow;
+
+	// p_hash[i] is the hash of the first i characters of the given string
+	vector<long long> p_hash;
+
+  public:
+	HashedString() {}
+	HashedString(const string &s) : p_hash(s.size() + 1) {
+		while (pow.size() <= s.size()) { pow.push_back((pow.back() * B) % M); }
+
+		p_hash[0] = 0;
+		for (int i = 0; i < (int)s.size(); i++) {
+			p_hash[i + 1] = ((p_hash[i] * B) % M + s[i]) % M;
+		}
+	}
+
+	long long get_hash(int start, int end) {
+		long long raw_val =
+		    (p_hash[end + 1] - (p_hash[start] * pow[end - start + 1]));
+		return (raw_val % M + M) % M;
+	}
+};
+vector<long long> HashedString::pow = {1};
+// EndCodeSnip
+
+int main() {
+	int test_num;
+	cin >> test_num;
+	for (int t = 0; t < test_num; t++) {
+		int n, k;
+		cin >> n >> k;
+		vector<string> s(n);
+		for (string &i : s) { cin >> i; }
+
+		vector<bool> res(n);
+		if (n < k) {
+			for (int i = 0; i < n; i++) {
+				for (int j = i + 1; j < n; j++) {
+					int diff_chars = 0;
+					for (int x = 0; x < k; x++) {
+						diff_chars += (s[i][x] != s[j][x]);
+					}
+					if (diff_chars <= 2) { res[i] = res[j] = true; }
+				}
+			}
+		} else {
+			vector<HashedString> hashed(n);
+			for (int i = 0; i < n; i++) { hashed[i] = HashedString(s[i]); }
```

Comment:
hm i think using `push_back` can remove the need
surely this code isn't that tight on time

Reply:
done